localstack/dns/server.py

```diff
@@ -0,0 +1,875 @@
+                    answer = [
+                        res.to_text() for answers in answer.response.answer for res in answers.items
+                    ]
+                except Timeout:
+                    answer = None
+                if not answer:
+                    continue
+                result = ns
+                break
+            except Exception:
+                pass
+
+        if result:
+            LOG.debug("Determined fallback dns: %s", result)
+        else:
+            LOG.info(
+                "Unable to determine fallback DNS. Please check if '%s' is reachable by your configured DNS servers"
+                "DNS fallback will be disabled.",
+                VERIFICATION_DOMAIN,
+            )
+        return result
+
+
+# ###### LEGACY METHODS ######
+def add_resolv_entry():
+    from localstack.services.edge import ensure_can_use_sudo
+
+    resolv_conf = "/etc/resolv.conf"
+    if os.path.exists(resolv_conf):
+        content = load_file(resolv_conf)
+
+        comment = "# The following line is required by LocalStack"
+        line = "nameserver 127.0.0.1"
+        if line not in content:
+            sudo_cmd = "" if is_root() else "sudo"
+            ensure_can_use_sudo()
+
+            for new_line in ("", line, comment):
+                # Surprisingly hard to find a reliable cross-platform shell
+                # solution to prepend text to a file, hence we use python here.
+                run(
+                    (
+                        '%s python -c "import sys; f=open(sys.argv[1]).read(); '
+                        "open(sys.argv[1], 'w').write('%s\\n' + f)\" %s"
+                    )
+                    % (sudo_cmd, new_line, resolv_conf)
+                )
+
+
+def setup_network_configuration():
+    # check if DNS is disabled
+    if not config.use_custom_dns():
+        return
+
+    # add entry to /etc/resolv.conf
+    if in_docker():
+        add_resolv_entry()
+
+
+def start_server(upstream_dns: str, port: int = config.DNS_PORT):
+    global DNS_SERVER
+
+    if DNS_SERVER:
+        # already started - bail
+        LOG.debug("DNS servers are already started. Avoid starting again.")
+        return
+
+    LOG.debug("Starting DNS servers (tcp/udp port %s on %s)..." % (port, config.DNS_ADDRESS))
+    dns_server = DnsServer(
+        port, protocols=["tcp", "udp"], host=config.DNS_ADDRESS, upstream_dns=upstream_dns
+    )
+    for name in NAME_PATTERNS_POINTING_TO_LOCALSTACK:
+        dns_server.add_host_pointing_to_localstack(name)
+    for skip_pattern in SKIP_PATTERNS:
+        dns_server.add_skip(skip_pattern)
+    if config.DNS_LOCAL_NAME_PATTERNS:
+        for skip_pattern in re.split(r"[,;\s]+", config.DNS_LOCAL_NAME_PATTERNS):
+            dns_server.add_skip(skip_pattern)
+    dns_server.start()
+    if not dns_server.wait_is_up(timeout=5):
+        LOG.warning("DNS server did not come up within 5 seconds.")
```

Comment:
In this case, we might want to do a `dns_server.shutdown()` to avoid leaking that server.